File: frontend/src/services/calendar.ts
/**
 * Calendar API Service
 */
import api from './api';

export interface CalendarEvent {
  id: number;
  event_id: number;
  title: string;
  description: string | null;
  event_date: string;
  duration: number;
  importance: number;
  owner_id: number;
  owner_name: string;
  owner_type: string;
  response: string | null;
}

export interface CalendarEventAttendee {
  id: number;
  character_id: number;
  event_response: string;
}

export const calendarService = {
  /**
   * List calendar events
   */
  listEvents: async (params?: {
    character_id?: number;
    from_date?: string;
    to_date?: string;
    response_filter?: string;
    limit?: number;
    offset?: number;
  }) => {
    const response = await api.get<CalendarEvent[]>('/calendar/', { params });
    return response.data;
  },

  /**
   * Get a specific event
   */
  getEvent: async (eventId: number) => {
    const response = await api.get<CalendarEvent>(`/calendar/${eventId}`);
    return response.data;
  },

  /**
   * Get event attendees
   */
  getEventAttendees: async (eventId: number) => {
    const response = await api.get<CalendarEventAttendee[]>(`/calendar/${eventId}/attendees`);
    return response.data;
  },

  /**
   * Respond to an event
   */
  respondToEvent: async (eventId: number, characterId: number, response: string) => {
    const apiResponse = await api.put(`/calendar/${eventId}/respond`, {
      character_id: characterId,
      response,
    });
    return apiResponse.data;
  },

  /**
   * Get upcoming events
   */
  getUpcomingEvents: async (characterId?: number, days: number = 7) => {
    const response = await api.get<CalendarEvent[]>('/calendar/upcoming/', {
      params: { character_id: characterId, days },
    });
    return response.data;
  },

  /**
   * Trigger calendar sync
   */
  triggerSync: async (characterId: number) => {
    const response = await api.post(`/calendar/sync/${characterId}`);
    return response.data;
  },
};

File: backend/app/models/calendar.py
"""
Calendar models

EVE Online calendar events system
"""
from sqlalchemy import Column, Integer, String, DateTime, Text, ForeignKey, BigInteger
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import relationship
from datetime import datetime, timezone
from app.core.database import Base


class CalendarEvent(Base):
    """Calendar events"""

    __tablename__ = "calendar_events"

    id = Column(Integer, primary_key=True, index=True)
    character_id = Column(Integer, ForeignKey("characters.id"), nullable=False, index=True)
    event_id = Column(BigInteger, unique=True, index=True)  # EVE event ID
    title = Column(String(255))
    description = Column(Text)
    event_date = Column(DateTime(timezone=True), index=True)
    duration = Column(Integer)  # Duration in minutes
    importance = Column(Integer)  # 0 = normal, 1 = important

    # Owner information
    owner_id = Column(BigInteger)
    owner_name = Column(String(255))
    owner_type = Column(String(50))  # character, corporation, alliance

    # Event response
    response = Column(String(50))  # accepted, declined, tentative, not_responded

    # Metadata
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))

    # Relationships
    character = relationship("Character", back_populates="calendar_events")

    def __repr__(self):
        return f"<CalendarEvent(event_id={self.event_id}, title='{self.title}', date={self.event_date})>"


class CalendarEventAttendee(Base):
    """Calendar event attendees"""

    __tablename__ = "calendar_event_attendees"

    id = Column(Integer, primary_key=True, index=True)
    event_id = Column(BigInteger, ForeignKey("calendar_events.event_id"), nullable=False, index=True)
    character_id = Column(BigInteger)  # Attendee character ID
    event_response = Column(String(50))  # accepted, declined, tentative, not_responded

    # Metadata
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))

    def __repr__(self):
        return f"<CalendarEventAttendee(event_id={self.event_id}, character_id={self.character_id}, response='{self.event_response}')>"
